### PR TITLE
Adding python-versions-runner workflow

### PR DESCRIPTION
**Pull request by shilovmaksim**
## Description
Adding python-versions runner workflow to have the ability run "Build Python package" workflows for multiple python versions simultaneously.

## Related workitem
https://github.com/actions/virtual-environments-internal/issues/3206

## Example run
https://github.com/shilovmaksim/python-versions/runs/5215603883?check_suite_focus=true

### CODE REVIEW

**Review by al-cheb — APPROVED** [review]

File: .github/workflows/python-versions-runner.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            "inputs" = @{
+              "VERSION" = "${{ matrix.version }}"
+              "PUBLISH_RELEASES" = "false"
+            }
+          } | ConvertTo-Json
+          $headers = @{
+            Authorization="Bearer ${{ secrets.PERSONAL_TOKEN }}"
+          }
+          Invoke-RestMethod -uri "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/workflows/python-builder.yml/dispatches" -method POST -headers $headers -body $payload
+          Start-Sleep -seconds 40
```
Comment (miketimofeev) [inline .github/workflows/python-versions-runner.yml:38]: Can we get rid of this and set up a cycle with retries? Setting such magic values seems unreliable

File: .github/workflows/python-versions-runner.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+jobs:
+  trigger_builds:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        version: ${{ fromJson(github.event.inputs.versions) }}
+    steps:
+      - run: |
```
[the remark this answers is not in the document]
Comment (shilovmaksim) [reply .github/workflows/python-versions-runner.yml]: Updated

File: .github/workflows/python-versions-runner.yml (hunk 1 above)
Comment (shilovmaksim) [reply .github/workflows/python-versions-runner.yml]: Updated

File: builders/python-versions-runner.ps1 [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    [switch] $PublishRelease
+)
+
+$summary = $Versions | ForEach-Object -Parallel { 
+    Import-Module "./builders/invoke-workflow.ps1"
+    Invoke-Workflow -Version $_ -PublishRelease $Using:PublishRelease
+}
+Write-Host "Results of triggered workflows:"
+$summary | Out-String
+if ($summary.Conclusion -contains "failure" -or $summary.Conclusion -contains "cancelled") {
```
Comment (al-cheb) [inline builders/python-versions-runner.ps1:29]: use -in operator

File: builders/invoke-workflow.ps1 [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    $result = [PSCustomObject]@{
+        Version = $version;
+        Conclusion = $workflowToCheck.conclusion;
+        Url = $workflowToCheck.html_url
+    }
+    if ($workflowToCheck.conclusion -ne "success") {
+        Write-Host "Triggered workflow for version '$($version)' completed with result '$($workflowToCheck.conclusion)'. Check logs: $($workflowToCheck.html_url)"
+        return $result
+        break
```
Comment (al-cheb) [inline builders/invoke-workflow.ps1:52]: Why do we need useless break?

File: builders/invoke-workflow.ps1 [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        $workflowToCheck = Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($workflowToCheck.id)"
+        Write-Host "Workflow run with Id: $($workflowToCheck.id) for version '$($version)' - status '$($workflowToCheck.status)'"
+    }
+    $result = [PSCustomObject]@{
+        Version = $version;
+        Conclusion = $workflowToCheck.conclusion;
+        Url = $workflowToCheck.html_url
+    }
+    if ($workflowToCheck.conclusion -ne "success") {
+        Write-Host "Triggered workflow for version '$($version)' completed with result '$($workflowToCheck.conclusion)'. Check logs: $($workflowToCheck.html_url)"
```
Comment (al-cheb) [inline builders/invoke-workflow.ps1:50]: remove $() for version

File: builders/invoke-workflow.ps1 [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                break
+            }
+        }
+
+        Start-Sleep -Seconds 120
+        $workflowToCheck = Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($workflowToCheck.id)"
+        Write-Host "Workflow run with Id: $($workflowToCheck.id) for version '$($version)' - status '$($workflowToCheck.status)'"
+    }
+    $result = [PSCustomObject]@{
+        Version = $version;
```
Comment (al-cheb) [inline builders/invoke-workflow.ps1:45]: remove ;

File: builders/invoke-workflow.ps1 [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        while (!$workflowToCheck) {
+            Start-Sleep -seconds 40
+            $workflowRuns = (Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs").workflow_runs | Where-Object {$_.status -like "*progress*" -and $_.id -ne $env:GITHUB_RUN_ID}
+            $workflowToCheck = $workflowRuns | Where-Object {
+                ((Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($_.id)/jobs").jobs.steps.name -like "*$version").Count -gt 0
+            }
+            $retries ++
+            if ($retries -gt 10) {
+                Write-Host "No workflow triggered for version $version or something went wrong with fetching its status"
+                $result = [PSCustomObject]@{
```
Comment (al-cheb) [inline builders/invoke-workflow.ps1:30]: Move outside while and at 44 update Conclusion and Url only

File: builders/invoke-workflow.ps1 [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    Invoke-RestMethod -uri "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/workflows/python-builder.yml/dispatches" -method POST -headers $headers -body $payload
+
+    while ($workflowToCheck.status -ne "completed") {
+        while (!$workflowToCheck) {
+            Start-Sleep -seconds 40
+            $workflowRuns = (Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs").workflow_runs | Where-Object {$_.status -like "*progress*" -and $_.id -ne $env:GITHUB_RUN_ID}
+            $workflowToCheck = $workflowRuns | Where-Object {
+                ((Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($_.id)/jobs").jobs.steps.name -like "*$version").Count -gt 0
+            }
+            $retries ++
```
Comment (al-cheb) [inline builders/invoke-workflow.ps1:27]: remove space

File: builders/invoke-workflow.ps1 [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    Invoke-RestMethod -uri "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/workflows/python-builder.yml/dispatches" -method POST -headers $headers -body $payload
+
+    while ($workflowToCheck.status -ne "completed") {
+        while (!$workflowToCheck) {
+            Start-Sleep -seconds 40
+            $workflowRuns = (Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs").workflow_runs | Where-Object {$_.status -like "*progress*" -and $_.id -ne $env:GITHUB_RUN_ID}
+            $workflowToCheck = $workflowRuns | Where-Object {
+                ((Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($_.id)/jobs").jobs.steps.name -like "*$version").Count -gt 0
```
Comment (al-cheb) [inline builders/invoke-workflow.ps1:25]: Why do we need to check count? Is like not enought?

File: builders/python-versions-runner.ps1 [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+param(
+    [Parameter (Mandatory=$true, HelpMessage="Python version to trigger build for")]
+    [array] $Versions,
+    [Parameter (Mandatory=$false, HelpMessage="Whether to publish release for built version")]
+    [switch] $PublishRelease
+)
+
+$summary = $Versions | ForEach-Object -Parallel { 
+    Import-Module "./builders/invoke-workflow.ps1"
```
[the remark this answers is not in the document]
Comment (shilovmaksim) [reply builders/python-versions-runner.ps1]: Renamed

File: builders/invoke-workflow.ps1 [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        Version = $version;
+        Conclusion = $workflowToCheck.conclusion;
+        Url = $workflowToCheck.html_url
+    }
+    if ($workflowToCheck.conclusion -ne "success") {
+        Write-Host "Triggered workflow for version '$($version)' completed with result '$($workflowToCheck.conclusion)'. Check logs: $($workflowToCheck.html_url)"
+        return $result
+        break
+    }
+    Write-Host "Triggered workflow for version '$($version)' succeeded; Url: $($workflowToCheck.html_url)"
```
[the remark this answers is not in the document]
Comment (shilovmaksim) [reply builders/invoke-workflow.ps1]: Done

File: builders/invoke-workflow.ps1 (hunk 4 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.ps1]: Indeed

File: builders/invoke-workflow.ps1 (hunk 5 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.ps1]: Done

File: builders/invoke-workflow.ps1 (hunk 6 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.ps1]: Updated

File: builders/invoke-workflow.ps1 (hunk 8 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.ps1]: Done

File: builders/invoke-workflow.ps1 (hunk 9 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.ps1]: Indeed

File: builders/invoke-workflow.ps1 (hunk 7 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.ps1]: Done

File: builders/invoke-workflow.psm1 [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    while ($workflowToCheck.status -ne "completed") {
+        while (!$workflowToCheck) {
+            Start-Sleep -seconds 40
+            $workflowRuns = (Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs").workflow_runs | Where-Object {$_.status -like "*progress*" -and $_.id -ne $env:GITHUB_RUN_ID}
+            $workflowToCheck = $workflowRuns | Where-Object {
+                (Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($_.id)/jobs").jobs.steps.name -like "*$version"
+            }
+            $retries++
+            if ($retries -gt 10) {
+                Write-Host "No workflow triggered for version $version or something went wrong with fetching its status"
```
Comment (miketimofeev) [inline builders/invoke-workflow.psm1:35]:
```suggestion
                Write-Host "Workflow triggered for python version $version not found or something went wrong with fetching the workflow status"
```

File: builders/invoke-workflow.psm1 [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+                return $result
+            }
+        }
+
+        Start-Sleep -Seconds 120
+        $workflowToCheck = Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($workflowToCheck.id)"
+        Write-Host "Workflow run with Id: $($workflowToCheck.id) for version '$version' - status '$($workflowToCheck.status)'"
```
Comment (miketimofeev) [inline builders/invoke-workflow.psm1:42]:
```suggestion
        Write-Host "Workflow run with Id: $($workflowToCheck.id) for python version '$version' found - status '$($workflowToCheck.status)'"
```

File: builders/invoke-workflow.psm1 [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    $result.Conclusion = $workflowToCheck.conclusion
+    $result.Url = $workflowToCheck.html_url
+    if ($workflowToCheck.conclusion -ne "success") {
+        Write-Host "Triggered workflow for version '$version' completed with result '$($workflowToCheck.conclusion)'. Check logs: $workflowToCheck.html_url"
```
Comment (miketimofeev) [inline builders/invoke-workflow.psm1:47]:
```suggestion
        Write-Host "Triggered workflow for python version '$version' failed with result '$($workflowToCheck.conclusion)'. Check logs: $workflowToCheck.html_url"
```

File: builders/invoke-workflow.psm1 [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        return $result
+    }
+    Write-Host "Triggered workflow for version '$($version)' succeeded; Url: $($workflowToCheck.html_url)"
```
Comment (miketimofeev) [inline builders/invoke-workflow.psm1:50]:
```suggestion
    Write-Host "Triggered workflow for python version '$version' succeeded; Url: $($workflowToCheck.html_url)"
```

File: builders/invoke-workflow.psm1 [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    Invoke-RestMethod -uri "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/workflows/python-builder.yml/dispatches" -method POST -headers $headers -body $payload
+
+    $result = [PSCustomObject]@{
+        Version = $version
+        Conclusion = "failure"
+        Url = "Not run"
+    }
+        
+    while ($workflowToCheck.status -ne "completed") {
+        while (!$workflowToCheck) {
```
Comment (miketimofeev) [inline builders/invoke-workflow.psm1:27]: Can we split these two `while` loops as it looks more simple? Currently, `$workflowToCheck.status` will not exist when we check it for the first time
```pwsh
 while (-not $workflowToCheck) {
# wait for workflow to be triggered
}

while ($workflowToCheck.status -ne "completed") {
# check the workflow status
}
```

File: builders/invoke-workflow.psm1 (hunk 14 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.psm1:47]: Completed ≠ succeeded.
It can also be cancelled or skipped.

File: builders/invoke-workflow.psm1 (hunk 16 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.psm1]: Done

File: builders/invoke-workflow.psm1 (hunk 14 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.psm1]: Done

File: builders/invoke-workflow.psm1 [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+        "inputs" = @{
+            "VERSION" = "$Version"
+            "PUBLISH_RELEASES" = "$PublishRelease"
+        }
+    } | ConvertTo-Json
+    $headers = @{
+        Authorization="Bearer $env:PERSONAL_TOKEN"
+    }
+
+    Invoke-RestMethod -uri "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/workflows/python-builder.yml/dispatches" -method POST -headers $headers -body $payload
```
Comment (al-cheb) [inline builders/invoke-workflow.psm1:18]: What do you think to create a separate variable for uri part?

`$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions`

File: builders/invoke-workflow.psm1 [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    # Waiting for workflow to complete
+    while ($workflowToCheck.status -ne "completed") {
+        Start-Sleep -Seconds 120
+        $workflowToCheck = Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($workflowToCheck.id)"
+        Write-Host "Workflow run with Id: $($workflowToCheck.id) for version '$version' - status '$($workflowToCheck.status)'"
+    }
+    $result.Conclusion = $workflowToCheck.conclusion
+    $result.Url = $workflowToCheck.html_url
+    if ($workflowToCheck.conclusion -ne "success") {
+        Write-Host "Triggered workflow for version '$version' completed unsuccessfully with result '$($workflowToCheck.conclusion)'. Check the logs: $workflowToCheck.html_url"
```
Comment (al-cheb) [inline builders/invoke-workflow.psm1:47]: $workflowToCheck.html_url -> $($workflowToCheck.html_url)

File: builders/invoke-workflow.psm1 [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    } | ConvertTo-Json
+    $headers = @{
+        Authorization="Bearer $env:PERSONAL_TOKEN"
+    }
+
+    Invoke-RestMethod -uri "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/workflows/python-builder.yml/dispatches" -method POST -headers $headers -body $payload
+
+    $result = [PSCustomObject]@{
+        Version = $version
```
Comment (al-cheb) [inline builders/invoke-workflow.psm1:21]: nitpick: $version -> $Version

File: builders/invoke-workflow.psm1 (hunk 17 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.psm1]: Move to separate variable

File: builders/invoke-workflow.psm1 [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+        $workflowToCheck = Invoke-RestMethod "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/runs/$($workflowToCheck.id)"
+        Write-Host "Workflow run with Id: $($workflowToCheck.id) for version '$version' - status '$($workflowToCheck.status)'"
+    }
+    $result.Conclusion = $workflowToCheck.conclusion
+    $result.Url = $workflowToCheck.html_url
+    if ($workflowToCheck.conclusion -ne "success") {
+        Write-Host "Triggered workflow for version '$version' completed unsuccessfully with result '$($workflowToCheck.conclusion)'. Check the logs: $workflowToCheck.html_url"
+        return $result
+    }
+    Write-Host "Triggered workflow for version '$($version)' succeeded; Url: $($workflowToCheck.html_url)"
```
[the remark this answers is not in the document]
Comment (shilovmaksim) [reply builders/invoke-workflow.psm1]: Done

File: builders/invoke-workflow.psm1 (hunk 19 above)
Comment (shilovmaksim) [reply builders/invoke-workflow.psm1]: Updated all occurrences

File: builders/invoke-workflow.psm1 [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+        "ref" = "main"
+        "inputs" = @{
+            "VERSION" = "$Version"
+            "PUBLISH_RELEASES" = "$PublishRelease"
+        }
+    } | ConvertTo-Json
+    $headers = @{
+        Authorization="Bearer $env:PERSONAL_TOKEN"
+    }
+    $uri = "$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions/workflows/python-builder.yml/dispatches"
```
Comment (al-cheb) [inline builders/invoke-workflow.psm1:17]: I am mean to replace all occurrences(28,30,41 - lines) `"$env:GITHUB_API_URL/repos/$env:GITHUB_REPOSITORY/actions` in the script to the variable.

Comment (shilovmaksim) [reply builders/invoke-workflow.psm1:17]: I see. Makes sense.

**Review by miketimofeev — APPROVED** [review]

File: .github/workflows/python-versions-runner.yml [hunk 22]
```diff
@@ -0,0 +1,31 @@
+name: Python versions runner
+on:
+  workflow_dispatch:
+    inputs:
+      versions:
+        description: 'Versions to build'
+        required: true
+        default: '","'
+      publish-releases:
```
Comment (MaksimZhukov) [inline .github/workflows/python-versions-runner.yml:9]: Looks like we don't use this input at all

Comment (shilovmaksim) [reply .github/workflows/python-versions-runner.yml:9]: Indeed

**Review by MaksimZhukov — APPROVED** [review]

**Pull request merged**